# Covariance Matrix of assets (solution)

In [4]:
import sys
!{sys.executable} -m pip install -r requirements.txt

You are using pip version 9.0.1, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


In [5]:
import numpy as np
import pandas as pd
import time
import os
import quiz_helper
import matplotlib.pyplot as plt

In [6]:
%matplotlib inline
plt.style.use('ggplot')
plt.rcParams['figure.figsize'] = (14, 8)

### data bundle

In [7]:
import os
import quiz_helper
from zipline.data import bundles

In [8]:
os.environ['ZIPLINE_ROOT'] = os.path.join(os.getcwd(), '..', '..','data','module_4_quizzes_eod')
ingest_func = bundles.csvdir.csvdir_equities(['daily'], quiz_helper.EOD_BUNDLE_NAME)
bundles.register(quiz_helper.EOD_BUNDLE_NAME, ingest_func)
print('Data Registered')

Data Registered


### Build pipeline engine

In [9]:
from zipline.pipeline import Pipeline
from zipline.pipeline.factors import AverageDollarVolume
from zipline.utils.calendars import get_calendar

universe = AverageDollarVolume(window_length=120).top(500) 
trading_calendar = get_calendar('NYSE') 
bundle_data = bundles.load(quiz_helper.EOD_BUNDLE_NAME)
engine = quiz_helper.build_pipeline_engine(bundle_data, trading_calendar)

### View Data¶
With the pipeline engine built, let's get the stocks at the end of the period in the universe we're using. We'll use these tickers to generate the returns data for the our risk model.

In [10]:
universe_end_date = pd.Timestamp('2016-01-05', tz='UTC')

universe_tickers = engine\
    .run_pipeline(
        Pipeline(screen=universe),
        universe_end_date,
        universe_end_date)\
    .index.get_level_values(1)\
    .values.tolist()
    
universe_tickers

[Equity(0 [A]),
 Equity(1 [AAL]),
 Equity(2 [AAP]),
 Equity(3 [AAPL]),
 Equity(4 [ABBV]),
 Equity(5 [ABC]),
 Equity(6 [ABT]),
 Equity(7 [ACN]),
 Equity(8 [ADBE]),
 Equity(9 [ADI]),
 Equity(10 [ADM]),
 Equity(11 [ADP]),
 Equity(12 [ADS]),
 Equity(13 [ADSK]),
 Equity(14 [AEE]),
 Equity(15 [AEP]),
 Equity(16 [AES]),
 Equity(17 [AET]),
 Equity(18 [AFL]),
 Equity(19 [AGN]),
 Equity(20 [AIG]),
 Equity(21 [AIV]),
 Equity(22 [AIZ]),
 Equity(23 [AJG]),
 Equity(24 [AKAM]),
 Equity(25 [ALB]),
 Equity(26 [ALGN]),
 Equity(27 [ALK]),
 Equity(28 [ALL]),
 Equity(29 [ALLE]),
 Equity(30 [ALXN]),
 Equity(31 [AMAT]),
 Equity(32 [AMD]),
 Equity(33 [AME]),
 Equity(34 [AMG]),
 Equity(35 [AMGN]),
 Equity(36 [AMP]),
 Equity(37 [AMT]),
 Equity(38 [AMZN]),
 Equity(39 [ANDV]),
 Equity(40 [ANSS]),
 Equity(41 [ANTM]),
 Equity(42 [AON]),
 Equity(43 [AOS]),
 Equity(44 [APA]),
 Equity(45 [APC]),
 Equity(46 [APD]),
 Equity(47 [APH]),
 Equity(48 [ARE]),
 Equity(49 [ARNC]),
 Equity(50 [ATVI]),
 Equity(51 [AVB]),
 Equity(

In [11]:
len(universe_tickers)

490

In [12]:
from zipline.data.data_portal import DataPortal

data_portal = DataPortal(
    bundle_data.asset_finder,
    trading_calendar=trading_calendar,
    first_trading_day=bundle_data.equity_daily_bar_reader.first_trading_day,
    equity_minute_reader=None,
    equity_daily_reader=bundle_data.equity_daily_bar_reader,
    adjustment_reader=bundle_data.adjustment_reader)

## Get pricing data helper function

In [13]:
from quiz_helper import get_pricing

## get pricing data into a dataframe

In [14]:
returns_df = \
    get_pricing(
        data_portal,
        trading_calendar,
        universe_tickers,
        universe_end_date - pd.DateOffset(years=5),
        universe_end_date)\
    .pct_change()[1:].fillna(0) #convert prices into returns

returns_df

Equity(0 [A])  Equity(1 [AAL])  Equity(2 [AAP])  \
2011-01-07 00:00:00+00:00       0.008437         0.014230         0.026702   
2011-01-10 00:00:00+00:00      -0.004174         0.006195         0.007435   
2011-01-11 00:00:00+00:00      -0.001886        -0.043644        -0.005927   
2011-01-12 00:00:00+00:00       0.017254        -0.008237         0.013387   
2011-01-13 00:00:00+00:00      -0.004559         0.000955         0.003031   
2011-01-14 00:00:00+00:00       0.003439        -0.009156         0.003022   
2011-01-18 00:00:00+00:00       0.034254        -0.062085        -0.004286   
2011-01-19 00:00:00+00:00      -0.010224        -0.008929         0.008754   
2011-01-20 00:00:00+00:00      -0.008496         0.021953        -0.004732   
2011-01-21 00:00:00+00:00       0.007873        -0.041038         0.005544   
2011-01-24 00:00:00+00:00       0.014646         0.027473        -0.001106   
2011-01-25 00:00:00+00:00      -0.006736         0.002982         0.009146   
2011-01-26 00:00:00+00:00      -0.030736         0.066133         0.003593   
2011-01-27 00:00:00+00:00       0.007721         0.023178        -0.001553   
2011-01-28 00:00:00+00:00      -0.018846        -0.080553        -0.000936   
2011-01-31 00:00:00+00:00       0.003608        -0.023615        -0.002351   
2011-02-01 00:00:00+00:00       0.011654        -0.001047        -0.009218   
2011-02-02 00:00:00+00:00       0.010112        -0.039304        -0.027477   
2011-02-03 00:00:00+00:00      -0.000289         0.007310         0.014126   
2011-02-04 00:00:00+00:00       0.005627        -0.036500         0.024014   
2011-02-07 00:00:00+00:00       0.007709         0.052046         0.008114   
2011-02-08 00:00:00+00:00       0.010854         0.016455         0.006202   
2011-02-09 00:00:00+00:00       0.004664         0.000000         0.016955   
2011-02-10 00:00:00+00:00       0.000413        -0.003048        -0.011367   
2011-02-11 00:00:00+00:00      -0.007150         0.028364         0.000764   
2011-02-14 00:00:00+00:00       0.001663        -0.015790        -0.023274   
2011-02-15 00:00:00+00:00      -0.011905         0.011043        -0.003926   
2011-02-16 00:00:00+00:00       0.015123         0.001958         0.013860   
2011-02-17 00:00:00+00:00      -0.003311        -0.017791        -0.024844   
2011-02-18 00:00:00+00:00       0.011078        -0.020103        -0.006693   
...                                  ...              ...              ...   
2015-11-20 00:00:00+00:00       0.001072        -0.002373         0.002767   
2015-11-23 00:00:00+00:00      -0.007094         0.002379        -0.001228   
2015-11-24 00:00:00+00:00       0.002085        -0.025309         0.003504   
2015-11-25 00:00:00+00:00      -0.008206         0.001938         0.006805   
2015-11-27 00:00:00+00:00      -0.003256         0.009201         0.003285   
2015-11-30 00:00:00+00:00      -0.010209        -0.010321        -0.012798   
2015-12-01 00:00:00+00:00       0.015748         0.048493        -0.002396   
2015-12-02 00:00:00+00:00      -0.005284         0.012930        -0.027166   
2015-12-03 00:00:00+00:00      -0.009521        -0.012555        -0.019944   
2015-12-04 00:00:00+00:00       0.020199         0.039303         0.006779   
2015-12-07 00:00:00+00:00      -0.000340         0.018020        -0.029578   
2015-12-08 00:00:00+00:00      -0.023307        -0.026876        -0.008866   
2015-12-09 00:00:00+00:00      -0.005170        -0.020213         0.022223   
2015-12-10 00:00:00+00:00       0.015039         0.010092        -0.009466   
2015-12-11 00:00:00+00:00      -0.012227        -0.045356        -0.019179   
2015-12-14 00:00:00+00:00      -0.011197        -0.007618        -0.008325   
2015-12-15 00:00:00+00:00       0.024683         0.019768         0.057990   
2015-12-16 00:00:00+00:00       0.010780         0.014190         0.029771   
2015-12-17 00:00:00+00:00      -0.017212        -0.017122        -0.048190   
2015-12-18 00:00:00+00:00      -0.041084        -

## Let's look at a two stock portfolio

Let's pretend we have a portfolio of two stocks.  We'll pick Apple and Microsoft in this example.

In [15]:
aapl_col = returns_df.columns[3]
msft_col = returns_df.columns[312]
asset_return_1 = returns_df[aapl_col].rename('asset_return_aapl')
asset_return_2 = returns_df[msft_col].rename('asset_return_msft')
asset_return_df = pd.concat([asset_return_1,asset_return_2],axis=1)
asset_return_df.head(2)

asset_return_aapl  asset_return_msft
2011-01-07 00:00:00+00:00           0.007146          -0.007597
2011-01-10 00:00:00+00:00           0.018852          -0.013311

## Factor returns
Let's make up a "factor" by taking an average of all stocks in our list.  You can think of this as an equal weighted index of the 490 stocks, kind of like a measure of the "market".  We'll also make another factor by calculating the median of all the stocks.  These are mainly intended to help us generate some data to work with.  We'll go into how some common risk factors are generated later in the lessons.

Also note that we're setting axis=1 so that we calculate a value for each time period (row) instead of one value for each column (assets).

In [16]:
factor_return_1 = returns_df.mean(axis=1)
factor_return_2 = returns_df.median(axis=1)
factor_return_l = [factor_return_1, factor_return_2]

## Factor exposures

Factor exposures refer to how "exposed" a stock is to each factor.  We'll get into this more later.  For now, just think of this as one number for each stock, for each of the factors.

In [17]:
from sklearn.linear_model import LinearRegression

In [18]:
"""
For now, just assume that we're calculating a number for each 
stock, for each factor, which represents how "exposed" each stock is
to each factor. 
We'll discuss how factor exposure is calculated later in the lessons.
"""
def get_factor_exposures(factor_return_l, asset_return):
    lr = LinearRegression()
    X = np.array(factor_return_l).T
    y = np.array(asset_return.values)
    lr.fit(X,y)
    return lr.coef_

In [19]:
factor_exposure_l = []
for i in range(len(asset_return_df.columns)):
    factor_exposure_l.append(
        get_factor_exposures(factor_return_l,
                             asset_return_df[asset_return_df.columns[i]]
                            ))
    
factor_exposure_a = np.array(factor_exposure_l)

In [20]:
print(f"factor_exposures for asset 1 {factor_exposure_a[0]}")
print(f"factor_exposures for asset 2 {factor_exposure_a[1]}")

factor_exposures for asset 1 [ 1.35101534 -0.58353198]
factor_exposures for asset 2 [-0.2283345   1.16364007]


## Quiz: Variance of stock 1

Calculate the variance of stock 1.  
$\textrm{Var}(f_{1}) = \beta_{1,1}^2 \textrm{Var}(f_{1}) + \beta_{1,2}^2 \textrm{Var}(f_{2}) + 2\beta_{1,1}\beta_{1,2}\textrm{Cov}(f_{1},f_{2}) + \textrm{Var}(s_{1})$

In [21]:
factor_exposure_1_1 = factor_exposure_a[0][0]
factor_exposure_1_2 = factor_exposure_a[0][1]
common_return_1 = factor_exposure_1_1 * factor_return_1 + factor_exposure_1_2 * factor_return_2
specific_return_1 = asset_return_1 - common_return_1

In [22]:
covm_f1_f2 = np.cov(factor_return_1,factor_return_2,ddof=1) #this calculates a covariance matrix
# TODO: get the variance of each factor, and covariances from the covariance matrix covm_f1_f2
var_f1 = covm_f1_f2[0,0]
var_f2 = covm_f1_f2[1,1]
cov_f1_f2 = covm_f1_f2[0][1]

# TODO: calculate the specific variance.  
var_s_1 = np.var(specific_return_1,ddof=1)

# TODO: calculate the variance of asset 1 in terms of the factors and specific variance
var_asset_1 = (factor_exposure_1_1**2 * var_f1) + \
              (factor_exposure_1_2**2 * var_f2) + \
              2 * (factor_exposure_1_1 * factor_exposure_1_2 * cov_f1_f2) + \
              var_s_1
print(f"variance of asset 1: {var_asset_1:.8f}")

variance of asset 1: 0.00028209


## Quiz 2: Variance of stock 2
Calculate the variance of stock 2.  
$\textrm{Var}(f_{2}) = \beta_{2,1}^2 \textrm{Var}(f_{1}) + \beta_{2,2}^2 \textrm{Var}(f_{2}) + 2\beta_{2,1}\beta_{2,2}\textrm{Cov}(f_{1},f_{2}) + \textrm{Var}(s_{2})$

In [23]:
factor_exposure_2_1 = factor_exposure_a[1][0]
factor_exposure_2_2 = factor_exposure_a[1][1]
common_return_2 = factor_exposure_2_1 * factor_return_1 + factor_exposure_2_2 * factor_return_2
specific_return_2 = asset_return_2 - common_return_2

In [24]:
# Notice we already calculated the variance and covariances of the factors

# TODO: calculate the specific variance of asset 2
var_s_2 = np.var(specific_return_2,ddof=1)

# TODO: calcualte the variance of asset 2 in terms of the factors and specific variance
var_asset_2 = (factor_exposure_2_1**2 * var_f1) + \
              (factor_exposure_2_2**2 * var_f2) + \
              (2 * factor_exposure_2_1 * factor_exposure_2_2 * cov_f1_f2) + \
              var_s_2
            
print(f"variance of asset 2: {var_asset_2:.8f}")

variance of asset 2: 0.00021856


## Quiz 3: Covariance of stocks 1 and 2
Calculate the covariance of stock 1 and 2.  
$\textrm{Cov}(f_{1},f_{2}) = \beta_{1,1}\beta_{2,1}\textrm{Var}(f_{1}) + \beta_{1,1}\beta_{2,2}\textrm{Cov}(f_{1},f_{2}) + \beta_{1,2}\beta_{2,1}\textrm{Cov}(f_{1},f_{2}) + \beta_{1,2}\beta_{2,2}\textrm{Var}(f_{2})$

In [25]:
# TODO: calculate the covariance of assets 1 and 2 in terms of the factors
cov_asset_1_2 = (factor_exposure_1_1 * factor_exposure_2_1 * var_f1) + \
            (factor_exposure_1_1 * factor_exposure_2_2 * cov_f1_f2) + \
            (factor_exposure_1_2 * factor_exposure_2_1 * cov_f1_f2) + \
            (factor_exposure_1_2 * factor_exposure_2_2 * var_f2)
print(f"covariance of assets 1 and 2: {cov_asset_1_2:.8f}")

covariance of assets 1 and 2: 0.00007133


## Quiz 4: Do it with Matrices!

Create matrices $\mathbf{F}$, $\mathbf{B}$ and $\mathbf{S}$, where  
$\mathbf{F}= \begin{pmatrix}
\textrm{Var}(f_1) & \textrm{Cov}(f_1,f_2) \\ 
\textrm{Cov}(f_2,f_1) & \textrm{Var}(f_2) 
\end{pmatrix}$
is the covariance matrix of factors,  

$\mathbf{B} = \begin{pmatrix}
\beta_{1,1}, \beta_{1,2}\\ 
\beta_{2,1}, \beta_{2,2}
\end{pmatrix}$ 
is the matrix of factor exposures, and  

$\mathbf{S} = \begin{pmatrix}
\textrm{Var}(s_i) & 0\\ 
0 & \textrm{Var}(s_j)
\end{pmatrix}$
is the matrix of specific variances.  

Then calculate $\mathbf{BFB}^T$ 

In [26]:
# TODO: covariance matrix of factors
F = covm_f1_f2
F

array([[1.02562520e-04, 9.79887017e-05],
       [9.79887017e-05, 9.44523986e-05]])

In [27]:
# TODO: matrix of factor exposures
B = factor_exposure_a
B

array([[ 1.35101534, -0.58353198],
       [-0.2283345 ,  1.16364007]])

#### Hint: for diagonal matrices
You can try using [numpy.diag](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.diag.html)

In [28]:
# TODO: matrix of specific variances
S = np.diag([var_s_1,var_s_2])
S

array([[0.00021723, 0.        ],
       [0.        , 0.00013739]])

In [29]:
# TODO: covariance matrix of assets
covm_assets = B.dot(F).dot(B.T) + S
print(f"covariance matrix of assets 1 and 2 \n{covm_assets}")

covariance matrix of assets 1 and 2 
[[2.82089786e-04 7.13296510e-05]
 [7.13296510e-05 2.18561378e-04]]
